hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/HoodieTableFactory.java

```diff
@@ -235,6 +235,11 @@ private static void setupHoodieKeyOptions(Configuration conf, CatalogTable table
       conf.setString(FlinkOptions.KEYGEN_CLASS_NAME, ComplexAvroKeyGenerator.class.getName());
       LOG.info("Table option [{}] is reset to {} because record key or partition path has two or more fields",
           FlinkOptions.KEYGEN_CLASS_NAME.key(), ComplexAvroKeyGenerator.class.getName());
+    } else if (!conf.getOptional(FlinkOptions.KEYGEN_CLASS_NAME).isPresent()) {
+      String keyGenName = FlinkOptions.getKeyGenClassNameByType(conf);
+      conf.setString(FlinkOptions.KEYGEN_CLASS_NAME, keyGenName);
+      LOG.info("Table option [{}] is reset to {} because of {}",
+          FlinkOptions.KEYGEN_CLASS_NAME.key(), keyGenName, FlinkOptions.KEYGEN_TYPE);
```

Comment:
Can we fix StreamerUtil line 218 instead ? Just give a default value `SimpleAvroKeyGenerator` should be enough, and this is a table config, the write config key gen type could overwrite this.

Reply:
Do you mean StreamerUtil line 318?
like :

```java
    String keyGeneratorClassName = conf.getString(FlinkOptions.KEYGEN_CLASS_NAME);
    if (keyGeneratorClassName == null) {
      keyGeneratorClassName = FlinkOptions.getKeyGenClassNameByType(conf);
    }

    HoodieTableMetaClient metaClient = HoodieTableMetaClient.withPropertyBuilder()
         .setKeyGeneratorClassProp(keyGeneratorClassName)
```

Reply:
Just like this:
```java
.setKeyGeneratorClassProp(conf.getOptional(FlinkOptions.KEYGEN_CLASS_NAME).orElse(SimpleAvroKeyGenerator.class.getName()))
```

Reply:
Here is the logic to create the real KeyGenerator   `HoodieAvroKeyGeneratorFactory#createKeyGenerator `:
https://github.com/apache/hudi/blob/75f6266594245cee998a5530711f929023d87a44/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/factory/HoodieAvroKeyGeneratorFactory.java#L51

If KEYGENERATOR_CLASS_NAME is not configured, it will be created according to KEYGENERATOR_TYPE.

So, I'm afraid there will be an error here:
If we use your modification method above, assuming that the user sets the type to complex, the class saved in the table properties is `SimpleAvroKeyGenerator`, but the `ComplexAvroKeyGenerator` is actually used. Used and saved are not the same.

Reply:
> Just give a default value SimpleAvroKeyGenerator should be enough, and this is a table config, the write config key gen type could overwrite this.

I have explained the reason.